tests/components/sensor/test_teleinfo.py

```diff
@@ -0,0 +1,86 @@
+"""The tests for the teleinfo platform."""
+
+import json
+import unittest
+from unittest.mock import patch
+
+from homeassistant.setup import setup_component
+
+from tests.common import (get_test_home_assistant, load_fixture)
+
+
+VALID_CONFIG_MINIMAL = {
+    'sensor': {
+        'platform': 'teleinfo',
+        'device': '/dev/ttyUSB0',
+    }
+}
+
+VALID_CONFIG_NAME = {
+    'sensor': {
+        'platform': 'teleinfo',
+        'name': 'edf',
+        'device': '/dev/ttyUSB0',
+    }
+}
+
+
+class KylinMock():
+    """Mock class for the kylin.Kylin object."""
+
+    def __init__(self, port, timeout=0):
+        """Initialize Kylin bject."""
+        self.port = port
+        self.timeout = timeout
+        self.sample_data = bytes(load_fixture('teleinfo.txt'), 'ascii')
+
+    def open(self):
+        """Open serial connection."""
+        pass
+
+    def close(self):
+        """Close serial connection."""
+        pass
+
+    def readframe(self):
+        """Return sample values."""
+        return json.loads(self.sample_data.decode('ascii'))
+
+
+class TestTeleinfoSensor(unittest.TestCase):
+    """Test the teleinfo sensor."""
+
+    def setUp(self):
+        """Set up things to run when tests begin."""
+        self.hass = get_test_home_assistant()
+        self.config = VALID_CONFIG_NAME
+
+    def tearDown(self):
+        """Stop everything that was started."""
+        self.hass.stop()
+
+    def _check_teleinfo_values(self, state):
+        self.assertEqual(state.state, '0501340443287')
+        self.assertEqual(state.attributes.get('HCHC'), 28005279)
+        self.assertEqual(state.attributes.get('HCHP'), 50392600)
+        self.assertEqual(state.attributes.get('OPTARIF'), 'HC..')
+        self.assertEqual(state.attributes.get('ISOUSC'), 30)
+        self.assertEqual(state.attributes.get('IINST'), 1)
+        self.assertEqual(state.attributes.get('IMAX'), 32)
+        self.assertEqual(state.attributes.get('PAPP'), 330)
+
+    @patch('kylin.Kylin', new=KylinMock)
```

Comment:
Instead you can do
`@patch.dict('sys.modules', kylyn=KylinMock)`
Then your test won't depend on kylyn at all and you will be able to remove it from script/gen_requirements_all.py

Note that you will need to mock `kylyn` (including kylyn.exceptions that you use) and not just `kylyn.Kylyn`

Reply:
i try that : 

```python
def mocked_exception(*args, **kwargs):
    raise Exception()

    # @patch('kylin.Kylin', new=KylinMock)
    @patch.dict('sys.modules', kylyn=KylinMock)
    @patch('kylin.exceptions.KylinNotFoundError', mocked_exception)
    def test_teleinfo_with_minimal_configuration(self):
        """Test Teleinfo with minimal configuration."""
        assert setup_component(self.hass, 'sensor', VALID_CONFIG_MINIMAL)
        state = self.hass.states.get('sensor.teleinfo')
        self.assertEqual("Teleinfo", state.name)
        self._check_teleinfo_values(state)
```

but i've got this error now : 

```
------------------------------------------------------------------------------------------------ Captured log call -------------------------------------------------------------------------------------------------
selector_events.py          65 DEBUG    Using selector: EpollSelector
setup.py                   183 INFO     Setting up sensor
entity_component.py        149 INFO     Setting up sensor.teleinfo
teleinfo.py                144 INFO     Read teleinfo informations
__init__.py                131 INFO     Port: /dev/ttyACM2 - N5321 gw - N5321 gw Mobile Broadband GPS Port
__init__.py                131 INFO     Port: /dev/ttyACM1 - N5321 gw - N5321 gw Mobile Broadband Data Modem
__init__.py                131 INFO     Port: /dev/ttyACM0 - N5321 gw - N5321 gw Mobile Broadband Modem
entity_component.py        191 ERROR    Error while setting up platform teleinfo
Traceback (most recent call last):
  File "/home/nlamirault/Perso/home-assistant/homeassistant/helpers/entity_component.py", line 170, in _async_setup_platform
    SLOW_SETUP_MAX_WAIT, loop=self.hass.loop)
  File "/usr/lib64/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/lib64/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/nlamirault/Perso/home-assistant/homeassistant/components/sensor/teleinfo.py", line 63, in setup_platform
    if not teleinfo_data.update():
  File "/home/nlamirault/Perso/home-assistant/homeassistant/util/__init__.py", line 306, in wrapper
    result = method(*args, **kwargs)
  File "/home/nlamirault/Perso/home-assistant/homeassistant/components/sensor/teleinfo.py", line 146, in update
    self._teleinfo = kylin.Kylin(port=self._device, timeout=2)
  File "/home/nlamirault/Perso/home-assistant/.tox/py36/lib/python3.6/site-packages/kylin/__init__.py", line 41, in __init__
    raise exceptions.KylinNotFoundError(port)
  File "/home/nlamirault/Perso/home-assistant/tests/components/sensor/test_teleinfo.py", line 52, in mocked_exception
    raise Exception()
Exception
setup.py                   201 INFO     Setup of domain sensor took 0.0 seconds.
core.py                    406 INFO     Bus:Handling <Event component_loaded[L]: component=sensor>
core.py                    406 INFO     Bus:Handling <Event homeassistant_stop[L]>
core.py                    406 INFO     Bus:Handling <Event homeassistant_close[L]>

 tests/components/sensor/test_teleinfo.py ⨯                                                                                                                                                          100% ██████████
```

I'm not sure Kylin is mocked.

Reply:
Oh, it's my typo: kylin and not kylyn

Reply:
@nlamirault can you try it out with the typo by Andrey corrected?